pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -206,13 +232,91 @@ func (c *Controller) doClusterBackupSync(cluster *v3.Cluster) error {
 	if time.Since(getBackupCompletedTime(newestBackup)) > backupIntervalHours {
 		newBackup, err := c.createNewBackup(cluster)
 		if err != nil {
-			return fmt.Errorf("[etcd-backup] error while create new backup for cluster [%s]: %v", cluster.Name, err)
+			return fmt.Errorf("error while create new backup for cluster [%s]: %v", cluster.Name, err)
 		}
-		logrus.Debugf("[etcd-backup] new backup created: %s", newBackup.Name)
+		log.Debugf("[etcd-backup] new backup created: %s", newBackup.Name)
 	}
 
-	// rotate old backups
-	return c.rotateExpiredBackups(cluster, clusterBackups)
+	return c.rotateExpiredBackups(cluster, recurringBackups)
+}
+
+func (c *Controller) doBackupForCluster(b *v3.EtcdBackup, cluster *v3.Cluster) (runtime.Object, error) {
+	var err error
+	if rketypes.BackupConditionCreated.IsUnknown(b) {
+		b.Spec.Filename = generateBackupFilename(b.Name, cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig)
+		b.Spec.BackupConfig = *cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig
+		rketypes.BackupConditionCreated.True(b)
+		// we set ConditionCompleted to Unknown to avoid incorrect "active" state
+		rketypes.BackupConditionCompleted.Unknown(b)
+		b, err = c.backupClient.Update(b)
+		if err != nil {
+			return b, err
+		}
+	}
+	bObj, saveErr := c.etcdSaveWithBackoff(b)
+	b, err = c.backupClient.Update(bObj.(*v3.EtcdBackup))
+	if err != nil {
+		return b, err
+	}
+
+	if saveErr != nil {
+		return b, fmt.Errorf("failed to perform etcd backup: %v", saveErr)
+	}
+	return b, nil
+}
+
+func anyBackupsRunning(cluster *v3.Cluster, backups []*v3.EtcdBackup) bool {
+	clusterTimeout := getTimeout(cluster)
+	for _, backup := range backups {
+		if !rketypes.BackupConditionCreated.IsTrue(backup) {
+			continue
+		}
+		backupCreationTime, err := getBackupCreatedTime(backup)
+		if err != nil {
+			log.Errorf("[etcd-backup] cluster [%s] backup [%s] is missing creation time: %v", cluster.Name, backup.Name, err)
+			continue
+		}
+		// cluster backup is younger than its timeout and completion is unknown
+		// therefore, it's currently running
+		if time.Since(backupCreationTime) < clusterTimeout && rketypes.BackupConditionCompleted.IsUnknown(backup) {
```

Comment:
A backup could be in any state if it's unknown. Do we have a way for the user to tell what state a backup is in if it's unknown, other than logs? Just a thought

Reply:
We may already have that and I just missed it!

Reply:
This actually tripped me up for a bit, because a condition has to exist and be set to `"Unknown"` for `IsUnknown` to be true from a [norman perspective](https://github.com/rancher/norman/blob/4feb41eafabded527f156460e70d77947f9a5973/condition/condition.go#L66), which is apparently against the [kubernetes community api conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#:~:text=Condition%20status%20values%20may%20be%20True%2C%20False%2C%20or%20Unknown.%20The%20absence%20of%20a%20condition%20should%20be%20interpreted%20the%20same%20as%20Unknown.%20How%20controllers%20handle%20Unknown%20depends%20on%20the%20Condition%20in%20question.). So in lieu of a running condition, setting Completed to Unknown means creation succeeded, and we are not sure if it's successful or failed, we just know that it is running. I think this is probably a bad approach in hindsight, as currently etcd backups only utilize the 3 following states: Creating, Activating, and Active.
![queue-4](https://user-images.githubusercontent.com/33796120/151892391-6930aeca-2bb5-48e5-9cf7-beb8bf5e9ba5.png)
I think another condition would be better in order to signify that the backups is currently running (I'm almost certain such a condition already exists, just have to figure out which).

Reply:
@thedadams Do you have any input regarding this? Specifically regarding what to do to make the etcd backup indicate it is progressing from a UI/UX perspective?

Reply:
Had a discussion and like @jakefhyde said, the [Norman frontend](https://github.com/rancher/norman/blob/4feb41eafabded527f156460e70d77947f9a5973/condition/condition.go#L66) uses a series of conditions (in other places as well) including Condition.Unknown to indicate a backup is currently running.

My initial concern was: There's no reason to change it unless it could potentially cause a slew of user issues. You know, if a backup fails while it's in 'Unknown' state it won't alert the user until the backup times out correct? It wouldn't be a bug no but it could be an inconvenience if a user has an annoyingly high timeout value set, and then are wondering what happened to their _running_ backup..

So Norman already has seton conditions that wouldn't be ideal to change because it's used in other places in addition to etcd backup, @thedadams correct me if I'm wrong there. The backend conditions just aren't intuitive when you look at them on the backend. etcdbackup.go https://github.com/rancher/rancher/blob/6ade3e8cf06e618d98b1b390c579755504fad2b2/pkg/controllers/management/etcdbackup/etcdbackup.go#L358-L362 does check to see whether a backup failed before timeout and if so, puts it into a Condition.Failed state. This is fine with me

Reply:
Ah I think there may be a disconnect here: If a backup fails, it fails instantly (i.e. if the minio endpoint doesn't exist anymore, just exit). This will immediately update the condition to False, and show correctly in the UI. The timeout only comes into play when the backup is taking too long (which is a pretty rare case, most etcd backups complete pretty quickly, however a minio upload on a slow network can cause this), where we consider the backup to be a failure if it goes longer than the timeout, at which point it fails.